src/daemon/screenlock.vala

```diff
@@ -0,0 +1,237 @@
+/*
+ * This file is part of budgie-desktop
+ *
+ * Copyright Budgie Desktop Developers
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ */
+
+namespace Budgie {
+    const string DBUS_SCREENLOCK = "org.buddiesofbudgie.BudgieScreenlock";
+	const string DBUS_SCREENLOCK_PATH = "/org/buddiesofbudgie/Screenlock";
+
+	[DBus (name="org.buddiesofbudgie.BudgieScreenlock")]
+    public class Screenlock {
+        static Screenlock? instance;
+
+        // connections to various schemas used in screenlocking
+        private GLib.Settings power;
+        private GLib.Settings session;
+        private GLib.Settings screensaver;
+        private GLib.Settings lockdown;
+
+        // public functions should check this variable first and exit gracefully where needed
+        bool all_apps = true;
+
+        private string calculate_lockcommand() {
+            // grab the background picture and strip the file prefix
+            string current_wallpaper = screensaver.get_string("picture-uri").replace("file://","");
+
+            // swaylock interprets : as a delimiter for the output name
+            // recommendation by swaylock is to replace : with ::
+            current_wallpaper = current_wallpaper.replace(":", "::");
+
+            string output = "";
+            File file = File.new_for_path(current_wallpaper);
+            if (current_wallpaper == "" || !file.query_exists()) {
+                output = "-c 000000";
+            }
+            else {
+                output = "-i " + current_wallpaper;
+            }
+
+            return "swaylock -Fklf " + output;
+        }
+
+        private string calculate_sleep() {
+            string output = "";
+
+            // number of seconds on ac power that is inactive before action taken; 0 is never
+            int sleep_inactive_ac_timeout = this.power.get_int("sleep-inactive-ac-timeout");
+
+            // BCC sets "nothing" or "suspend" when toggled
+            string sleep_inactive_ac_type = this.power.get_string("sleep-inactive-ac-type");
+
+            // number of seconds on battery that is inactive before action taken; 0 is never
+            int sleep_inactive_battery_timeout = this.power.get_int("sleep-inactive-battery-timeout");
+
+            // BCC sets "nothing" or "suspend" when toggled
+            string sleep_inactive_battery_type = this.power.get_string("sleep-inactive-battery-type");
+
+            if (sleep_inactive_battery_type == "suspend" && sleep_inactive_battery_timeout != 0) {
+                output = "timeout " + sleep_inactive_battery_timeout.to_string();
+                output += " 'if upower -i $(upower -e | grep '/battery') | grep --color=never -E 'state' | grep -e 'discharging'; then systemctl suspend; fi' ";
```

Comment:
Related to the upower code, I think we should instead be creating a Up.Client and listening in on the device changes (based on the PowerIndicator.vala code at least seems like changed events call device_added), where we get the DeviceState and determine if the device state is DeviceState.DISCHARGING.

At that point, we would call swayidle if that exists on the system, otherwise use the logind interface similar to UserIndicatorWindow.vala to suspend without interactivity.

Thoughts?

Reply:
fair enough.  Noted though that PowerIndicator is now using various deprecated APIs.